### DISCUSSION TITLE
A "set_trainable" method for VectorXvar/Variable/Expr

### DISCUSSION THREAD

**Discussion by f-izzat** [autodiff/autodiff#218] [Q&A] [answered]
I have the following class that inherits from VectorXvar.

```
    template<typename T>
    class Parameter;

    template<>
    class Parameter<Eigen::VectorXd> : public VectorXvar {
    bool trainable = true;
    public:
        Parameter() : VectorXvar() {}
        Parameter(const Eigen::VectorXd& val) : VectorXvar(val) {}

        void fix() {
            trainable = false;
        }
        void fix(const Eigen::Index& idx) {
            // Index the vector and set the expression to constant?
        }

        Parameter& operator=(const Parameter& oParam) {
            if (!trainable_) { throw std::runtime_error("Error fixed value.."); }
        }

        void unfix() {
            trainable = true;
        }

    };
```
Essentially, I would like to make all or some elements in the Vector "trainable" (analogous to tensorflows Variable, trainable option). What I was thinking was to change the expression to a `ConstantExpr` for some/all elements. Meaning, during an optimization process a specific element (or all) in the vector would remain unchanged in its value. As well as whenever its value is attempted to be changed (i.e. `operator=` is called) it would throw a `runtime_error` indicating that the `.fix()` method has been called previously

Have I understood what autodiff's underlying code does properly?
Should I instead inherit from `Expr` and add the `fix()` method there? If so how should I alter for the case of VectorXvar?
Or should I instead, create a new lone `Parameter` class that accepts an Eigen::VectorXd to which in a constructor I convert this to a VectorXvar?
Has there been an implementation like this previously?

Note:
The reason as to why I inherited in the first place is because in my equations I would perform several operations on `Parameter` directly.
 
Thanks

**Comment by allanleal** [comment, marked as answer]
Hi @f-izzat , I'm not familiar with TensorFlow and the trainable concept you mention above. But I think I understand what you want to do. I would try to find a way to do this avoiding the check (`!trainable`) for performance reasons. Unfortunately, I don't have time now to advise you further.
